### ISSUE TITLE
Implement Offsets for Expanded Frames

### ISSUE THREAD

**Issue by jrconway3** [bluecarrot16/lpctools#1]
So this is in response to the comment over at OGA. Figured I could drop it in here as a reminder.

Need to figure out the offsets for the "expanded" frames. As new frames get added it'd be ideal to get those offsets, too.

I think the main issue here would be in regards to the head position--that is what's standard here, and this tool is primarily useful for positioning assets on the head and neck (body assets may vary heavily, so that's probably less useful by comparison).

I'll grab lpctools and look into how it handles offsets, and figure out how to get the new offsets. This probably won't be a quick thing, but I'll try getting it done by the end of the week because I do have all week off. It'd be easier to get it done this week.

**Comment by jrconway3** [comment]
Sorry for the delay. Working on a preliminary. I'm still not sure exactly where the offsets come from... I see references to each animation type but nothing that says "+X, +Y" position or anything.

It almost looks like, based on the code, the offsets are actually from the images themselves? Well, I'm adding all references I can to the other frames anyway. If I can't figure out how to add the offsets directly, I'll just add the animation frames, the expanded layout, and then drop in here what the offsets actually are.

**Comment by jrconway3** [comment]
```
reference_points_male.png
```
Oh, so its literally just the top 1px of the head. Okay.


Maybe when I tried using arrange before I didn't catch this and that's what messed me up. Anyway, I can work on building this out now.

**Comment by jrconway3** [comment]
https://github.com/jrconway3/lpctools/blob/issue-1-add-new-animations/tests/arrange_files/hair/reference_points_universal_expanded.png

WIP... still gotta run some tests. I'll try it out a bit later.

**Comment by bluecarrot16** [comment]
Hi, sorry for the delay. I've been away for the holidays.

You're exactly right; the image offsets for the `arrange distribute` verb (and related verbs, e.g. `arrange distribute-repack` which have the same internals) are provided in an image via the `--offsets` argument.  They are not hard-coded in the code, and the offsets file you mentioned is just one example. I've used different offsets for different applications, compare e.g. the shields example to the hair example.  You can read about this by running `lpctools arrange distribute--help`  or looking here https://github.com/bluecarrot16/lpctools/blob/main/lpctools/__init__.py#L562-L567 . I realize now that this same verbiage isn't printed if you run `lpctools arrange distribute-repack --help`, and it probably should be. 

The offsets image tells lpctools where the center of the images will be placed, so a no-op offsets image would have a pixel in the center (e.g. 32,32 for a 64x64-px frame size). So yeah, all that's needed is to adapt this image to the additional frames. Hope this helps.

**Comment by jrconway3** [comment]
Okay, I went ahead and updated the reference sheets and added masks. I had to also clean up backslash/halfslash in the process because when I went to create the masks (select transparency, inverse selection) it found all the bad pixels I thought I removed before. So I grab all those out and I'll fix those sheets over at the ULPC Generator.


Anyway, I'll run tests on this soon to confirm its working. I'll start by generating new hairstyles to handle this. I also had to adjust the layout slightly; Sanderfrenken opted to split all types of assets into separate rows, whereas I was trying to save room by leaving idle/combat idle and sit/emotes in the same rows. (It doesn't matter much when we have the animations split anyway, so its not a big deal.)

My hope is that we can use the offsets and masks to generate the full sheet, then arrange and/or repack to split it into individual sheets. That would also give me the opportunity to easily add in grab/carry later.

**Comment by jrconway3** [comment]
`python bin/lpctools arrange distribute --layout universal-expanded --input tests/arrange_files/hair/hair_plain/ --output tests/arrange_files/hair/hair_plain_male.png --offsets tests/arrange_files/hair/reference_points_universal_expanded.png`

Generating a blank file...


Nevermind, even existing commands generate a blank file for me:
`python bin/lpctools arrange distribute --input tests/arrange_files/hair/hair_shoulderr/ --output tests/arrange_files/hair/hair_shoulder_r.png --offsets tests/arrange_files/hair/reference_points_male.png --mask tests/arrange_files/hair/masks_male.png`


This command doesn't work for me:
	```bash
	lpctools arrange distribute \
	--input tests/arrange_files/hair/hair_plain/ \
	--output tests/arrange_files/shield/hair_plain.png \
	--input tests/arrange_files/hair/hair_page2/ \
	--output tests/arrange_files/shield/hair_page2.png \
	--input tests/arrange_files/hair/hair_shoulderr/ \
	--output tests/arrange_files/shield/hair_shoulderr.png \
	--offsets tests/arrange_files/hair/reference_points_male.png \
	--mask tests/arrange_files/hair/masks_male.png
	```

I have to use "python" rather than "bash". And running lpctools from the main directory doesn't work, I have to use bin/lpctools. When I do this, it still generates blank files.


However, when I run recolors it works:
```
lpctools colors -v recolor --input tests/recolor_files/hair.png tests/recolor_files/hair_page2.png --mapping tests/recolor_files/palettes.json
```

Though `hair.png` does not exist, only `hair_plain.png` but when I swapped to the correct file, it worked.

**Comment by bluecarrot16** [comment]
> This command doesn't work for me: `bash lpctools arrange distribute \ --input tests/arrange_files/hair/hair_plain/ \ --output tests/arrange_files/shield/hair_plain.png \ --input tests/arrange_files/hair/hair_page2/ \ --output tests/arrange_files/shield/hair_page2.png \ --input tests/arrange_files/hair/hair_shoulderr/ \ --output tests/arrange_files/shield/hair_shoulderr.png \ --offsets tests/arrange_files/hair/reference_points_male.png \ --mask tests/arrange_files/hair/masks_male.png `
> 
> I have to use "python" rather than "bash". And running lpctools from the main directory doesn't work, I have to use bin/lpctools. When I do this, it still generates blank files.
> 

`bin/lpctools` is a Python program which serves as an entry point to the `lpctools` Python module. This is why `python bin/lpctools` does anything; however, you should be able to run the bare `lpctools` command (see below). You should not run `bash bin/lpctools` since `bin/lpctools` is not a bash script.

Did you install the required dependencies  using `conda` as described in the README (including the `pip install -e .` command)?

After installing, in each shell session you'll need to run `conda activate lpctools` before the `lpctools` command will work. This is because the installation instructions install the `lpctools` package (and all its dependencies) within a [Conda environment](https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/environments.html), isolating them from other Python packages, etc. When you `conda activate` an environment, it adds the executable(s) (`bin/lpctools`) to your `$PATH` and adds the necessary python packages to your `$PYTHONPATH` for the duration of that shell session. I'll add a note to that effect to the README.

Let me know if this step fixes the other problems you are having.

**Comment by jrconway3** [comment]
Yes I am running conda activate lpctools. I always run that.


Running lpctools directly, however, ALWAYS shows me this:
![Image](https://github.com/user-attachments/assets/df133016-b3e3-4e46-b811-eec0b3c49720)


As such, the only way I was able to get it to work was using python.exe bin/lpctools. (To make things easier, I copied lpctools to another directory I commonly use.)

I adjusted the angle of the window so you can see I _am_ running conda activate lpctools. Previously I had the anaconda window to the right and the popup blocked the left side of the window.


Other things I can't get working, for example, are this:
```
lpctools colors recolor \
		--input Universal-LPC-Spritesheet-Character-Generator/spritesheets/hair/male/*.png \
		--mapping tests/recolor_files/all-palettes.json
```

Wildcards don't work for me. I have to do the direct file, so I can't just generate a bunch of recolors at once. Makes things a bit more annoying now that the animations are split up.


I suspect the issue might be windows-related.

**Comment by bluecarrot16** [comment]
Yes, this is likely a Windows problem. Unfortunately I do not have access to a Windows machine and don't have enough knowledge of Windows to troubleshoot... I don't know if the syntax for invoking command line applications using Windows PowerShell is even the same as the example you posted (which was written for sh/bash)... In particular, expansion of that wildcard `*` is a shell feature, not an lpctools feature.

You'll probably have best luck running this by installing `conda` and `lpctools` into WSL...

**Comment by jrconway3** [comment]
All right. I'll look into WSL.

**Comment by jrconway3** [comment]
https://raw.githubusercontent.com/bluecarrot16/lpctools/a0da01fa277ce10e07a0377c05c1a24714d1918e/tests/arrange_files/hair/hair_page2.png

^_^

Ugh... finally got it working. Now I just need to run tests and make sure it aligns correctly. But then I also need to figure out how to properly align existing assets to arrange them...

**Comment by jrconway3** [comment]
Now I have WSL working to the point where conda activates as soon as I load WSL, so I don't have to remember the activate command anymore. I just need to run WSL in the console and I can jump straight in. This makes things SO MUCH EASIER.

These are the instructions I used for this:
https://kontext.tech/article/1064/install-miniconda-and-anaconda-on-wsl-2-or-linux


I should be able to run the tests a bit better now and confirm whether the alignments are working correctly. My bigger issue now is finding an easy way to line up these assets... I may need to create a template with the correct "center" positioning based on the hair positions here so I can line up future head assets easily.

**Comment by bluecarrot16** [comment]
> My bigger issue now is finding an easy way to line up these assets... I may need to create a template with the correct "center" positioning based on the hair positions here so I can line up future head assets easily.

Yes, it's a bit awkward because the offsets I commonly use correspond to the top-left-most pixel in the character's head, so your source images for the hair are not actually centered but are offset a bit down and to the right. Earlier scripts did it that way so I followed. 

As you mention, I use a template to align the source frame images for the hair: [hair-template.zip](https://github.com/user-attachments/files/18537255/hair-template.zip)

Align the hair assets to the sprites in this template. I use the combined sheets (`male.png` and `female.png` and use a function in my editor, PyxelEdit, to export the separate images (`e.png, `w.png`, etc.) to a directory. Then I use `lpctools arrange distribute` to make the combined sheets. See examples here <https://opengameart.org/content/lpc-hair> in the `_build` subdirectory .


To make new offsets, you can just draw a red line (or some color that's not in the hair ramp) at the appropriate position (corresponding to the top line of the character's head; really only the top-left pixel matters) over some hair asset you've already made the full sheet for. Then copy-paste that "annotated" version of the hair onto your full spritesheet. Then delete everything except the red line.

**Comment by jrconway3** [comment]
Funnily enough, I already had something similar here. I only have female though and I have all frames of "hurt" rather than 2-6. Anyway, I'll use this to create the basic sheets and see how easy it is to get it to map to all frames and then confirm it works on the full sheet as well.


Once I confirm it all works I'll open a PR.


EDIT: The offset I have in place is based on the one you have in tests. Its simply the top 1px line of the head. We don't even need  more than one head variant because heads are now supposed to all be in the exact same position. The asset might need more than one variant if we want to adjust around the shoulders or something, but that's probably better done manually anyway or have a second layer behind the body.

**Comment by jrconway3** [comment]
Nice, I managed to get it working. My first example of this in action is the updated Bascinet Round:
https://github.com/LiberatedPixelCup/Universal-LPC-Spritesheet-Character-Generator/pull/118


(The Bascinet Round w/Visor had a slightly different shape to the helmet. I melded the rounded back of it with the visor-less Bascinet, then used lpctools to drop it on all frames.)

**Comment by jrconway3** [comment]
I'm prioritizing getting this done now. As of right now it works quite well on its own, though I still keep forgetting to rename cast to spellcast when I split frames. That's the only notable issue I'm having at the moment.

This is working very well, I just want to write some tests out first:
https://github.com/jrconway3/lpctools/issues/1
